uh50_api/uh50.py

```diff
@@ -2,70 +2,300 @@
+    'return_temperature_max_previous_year_c': r'9.4\*01\(.*?\*C&(.*?)\*C',
+    'operating_hours': r'6.31\((.*?)\*h\)',
+    'fault_hours': r'6.32\((.*?)\*h\)',
+    'fault_hours_previous_year': r'6.32\*01\((.*?)\*h\)',
+    'yearly_set_day': r'6.36\((.*?)\)',
+    'monthly_set_day': r'6.36\*02\((.*?)\)',
+    'meter_date_time': r'9.36\((.*?)\)',
+    'measuring_range_m3ph': r'9.24\((.*?)\*m3ph\)',
+    'settings_and_firmware': r'9.1\((.*?)\)',
+    'flow_hours': r'9.31\((.*?)\*h\)',
+}
 
 def find_ports():
     "Returns the available ports"
     return serial.tools.list_ports.comports()
 
-def _search_data(data):
-    match = re.search( r'6.8\((.*)\*GJ\)6.26\((.*)\*m3\)9.21\(66153690\)', str(data), re.M|re.I)
-    if match: 
-        return match.group(1),match.group(2)
-
-    raise Exception("GJ and m3 values not found")
-
-def _connect_serial(port) -> Serial:
-    return serial.Serial(port,
-                        baudrate=300,
-                        bytesize=serial.SEVENBITS,
-                        parity=serial.PARITY_EVEN,
-                        stopbits=serial.STOPBITS_TWO,
-                        timeout=1,
-                        xonxoff=0,
-                        rtscts=0
-                        )
-
-def _wake_up(conn) -> str:
-    # Waking up should be done at 300 baud
-    # Sending /?!
-    conn.write(b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x2F\x3F\x21\x0D\x0A")
-    ir_command='/?!\x0D\x0A'
-    conn.write(ir_command.encode('utf-8'))
-    conn.flush()   
-    return conn.readline().decode('utf-8')[1:9]   # Read at 300 baud, this gives us the typenr
-
-def _get_data(conn):
-    #TODO make this asyc
-    conn.baudrate=2400 # Now switch to 2400 BAUD. This could be different for other models. Let me know if you experience problems.
-    return conn.readline().decode('utf-8') # Reading just one line, because that's where we know the data is.
+class UH50:
+    """Class for reading the UH50 on the specified port"""
+
+    def __init__(self, port: str) -> None:
+        self.port = port
+
+    def update(self) -> None:
+        "Reads the UH50 on the specified port, searching for info on GJ and m3"
+        with self._connect_serial() as conn:
+            self.model = self._wake_up(conn)
+
+            # checking if we can read the model (eg. 'LUGCUH50')
+            if not(self.model):
+                raise Exception('No model could be read')
+            
+            self._full_response = self._get_data(conn)
+            self._gj, self._m3 = self._search_data(self._full_response)
+
+    def update_dummy(self) -> None:
+        "Sets dummy values for testing purposes, when no live connection is available"
+        self.heat_usage_gj = '999.'+str(randrange(100,999)) 
+        self.volume_usage_m3 = '9999.'+str(randrange(10,99))
+        self.model = 'LUGCUH50' 
+
+    def _search_response(self, data):
```

Comment:
de _search_response bevat te veel herhalingals ik het patroontje goed begrijp, match je daar een regex op de searchstring en wil je het resultaat van die match in een aantal verschillende datatypes (int, float, str) terug. Als het parsen naar een int/float niet lukt, dan alsnog maar de string (gotta love dynamic typing)

Reply:
Dit zou een eigen class kunnen worden waarin het parsen van die response zit. Denk aan een class met een constructor waar die string in gaat en getters voor de fields die je nodig hebt. Daarin kan je dan een 3-tal private methodes gebruiken om een int/float/string terug te geven

Reply:
Dit heb ik nu werkend. Ik vind het wel even lastig dat ie nu live de transformatie doet in de getter ipv direct na de update/read, waarna het een data-object wordt die je evt ook als geheel kan aanbieden ergens. Nu krijg je de data immers alleen per aangeroepen functie.